src/gc.rs

```diff
@@ -0,0 +1,257 @@
+struct GcBox<T: ?Sized>(ManuallyDrop<T>);
+
+impl<T> GcBox<T> {
+    fn new(value: T) -> *mut GcBox<T> {
+        let (layout, _) = Layout::new::<usize>().extend(Layout::new::<T>()).unwrap();
+        let base_ptr = unsafe { GC_ALLOCATOR.alloc(layout).unwrap().0.as_ptr() };
+        let gcbox = GcBox(ManuallyDrop::new(value));
+        let obj_ptr = unsafe { (base_ptr as *mut usize).add(1) } as *mut GcBox<T>;
+
+        unsafe {
+            obj_ptr.copy_from_nonoverlapping(&gcbox, 1);
+            let fatptr: &mut dyn Finalizable = &mut *obj_ptr;
+            let mut vptr = transmute::<*const dyn Finalizable, (usize, usize)>(fatptr).1;
+            ::std::ptr::write(
+                base_ptr as *mut Option<NonZeroUsize>,
+                Some(NonZeroUsize::new_unchecked(vptr)),
+            );
+            ::std::ptr::write(base_ptr as *mut Option<NonZeroUsize>, None);
+        }
+
+        forget(gcbox);
+        let flzr = Self::finalizer_shim;
+        unsafe {
+            boehm::GC_register_finalizer(
+                base_ptr as *mut _ as *mut ::std::ffi::c_void,
+                flzr,
+                ::std::ptr::null_mut(),
+                ::std::ptr::null_mut(),
+                ::std::ptr::null_mut(),
+            );
+        }
+        obj_ptr
+    }
+
+    unsafe extern "C" fn finalizer_shim(obj: *mut c_void, meta: *mut c_void) {
+        let vptr = *(obj as *mut Option<NonZeroUsize>);
+        match vptr {
+            Some(nzptr) => {
+                let objptr = (obj as *mut usize).add(1); // UB I think?
+                let flzr = transmute::<(usize, usize), &mut dyn Finalizable>((
+                    objptr as usize,
+                    nzptr.get(),
+                ));
+                flzr.finalize()
+            }
+            None => return,
+        }
+    }
+
+    fn new_from_layout(layout: Layout) -> NonNull<GcBox<MaybeUninit<T>>> {
+        unsafe {
+            let (nl, _) = Layout::new::<usize>().extend(layout).unwrap();
+            let base_ptr = GC_ALLOCATOR.alloc(nl).unwrap().0.as_ptr() as *mut usize;
+
+            // Placeholder for vptr
+            ::std::ptr::write(base_ptr as *mut Option<NonZeroUsize>, None);
+
+            NonNull::new_unchecked((base_ptr.add(1)) as *mut GcBox<MaybeUninit<T>>)
+        }
+    }
+}
+
+impl<T> GcBox<MaybeUninit<T>> {
+    unsafe fn assume_init(&mut self) -> NonNull<GcBox<T>> {
+        // With T now considered initialized, we must make sure that if GcBox<T>
+        // is reclaimed, T will be dropped. We need to find its vptr and replace the
+        // GcDummyDrop vptr in the block header with it.
+        let fatptr: &mut dyn Finalizable = self;
+        let vptr = NonZeroUsize::new_unchecked(
+            transmute::<&mut dyn Finalizable, (usize, usize)>(fatptr).1,
+        );
+        let base_ptr = (self as *mut _ as *mut usize).sub(1) as *mut Option<NonZeroUsize>;
+        let flzr = Self::finalizer_shim;
+
+        ::std::ptr::write(base_ptr, Some(vptr));
```

Comment:
This chunk of code feels very similar to that at line 152: maybe it should be in its own function?

Reply:
I've simplified this quite a bit here: e53a221a81efe70620a0126699cf3ff0e196ebd4 See if you prefer it.

Reply:
Hmm, I'm not sure it's any better -- seems to be the same number of lines of code either way?

Reply:
Agreed on the number of lines, I do think it's marginally better though because it reduces some code duplication.

Reply:
OK.